### PR TITLE
Replace Logger.{Level,SetLevel} with DynamicLevel

### PR DESCRIPTION
**Pull request by jcorbin**
Depends on #172, supercedes #173.

After this PR, the only part of #166 left is the CheckedMessage composition work.

### CODE REVIEW

**Comment by jcorbin** [comment]
Rebased, ready for review

File: http_handler.go [hunk 1]
```diff
@@ -35,18 +35,19 @@ type errorResponse struct {
 }
 
 type levelHandler struct {
-	logger Logger
+	lvl AtomicLevel
```
Comment (akshayjshah) [inline http_handler.go]: Why not just add `ServeHTTP` to the `AtomicLevel` type? It doesn't add any overhead if it's not used, and it removes one type and wrapper from this now-crowded package.

File: level.go [hunk 2]
```diff
@@ -105,3 +121,41 @@ func (l *Level) UnmarshalText(text []byte) error {
 	}
 	return nil
 }
+
+// Enabled return true if the given level is at or above this level.
```
Comment (akshayjshah) [inline level.go]: return -> returns

File: level.go [hunk 3]
```diff
@@ -105,3 +121,41 @@ func (l *Level) UnmarshalText(text []byte) error {
...
+func (l Level) Enabled(lvl Level) bool {
+	return lvl >= l
+}
+
+// DynamicLevel creates an atomically changeable dynamic logging level.  The
+// returned level can be passed as a logger option just like a concrete level.
+//
+// The value's SetLevel() method may be called later to change the enabled
+// logging level of all loggers that were passed the value (either explicitly,
+// or by creating sub-loggers with Logger.With).
```
Comment (akshayjshah) [inline level.go]: nice - this is a much clearer explanation than the parent/child terminology I was using.

File: meta.go [hunk 4]
```diff
@@ -37,37 +33,25 @@ type Meta struct {
 	Hooks       []Hook
 	Output      WriteSyncer
 	ErrorOutput WriteSyncer
-
-	lvl *atomic.Int32
+	LevelEnabler
```
Comment (akshayjshah) [inline meta.go]: Can we keep embeds at the top, separated from the other fields with a blank line?

**Comment by akshayjshah** [comment]
Great - appreciate the quick turnaround after a delayed review. Feel free to merge after resolving the conflict.

**Comment by anuptalwalkar** [comment]
A quick question on this PR. You seem to have removed the Level access on Logger object. How will a user know what log level is set on the created/passed zap.Logger object?

**Comment by jcorbin** [comment]
@anuptalwalkar the idea now is that there isn't necessary a "the level of a logger object" concept; I'd respond to your question of "how?" with a clarifying "why?":
- if what you're trying to do is do your own filtering style logic, then you should be just calling `Logger.Check`
- otherwise, you may want to be using zap `zap.DynamicLevel`, which you can ask "what's the effective logging level"?

If those don't meet your needs, can you describe your use case a little more? (maybe in a new issue?)

**Comment by anuptalwalkar** [comment]
I see. my concerns were around setting up the Logger correctly. So now I can add `zap.DynamicLevel` as an option, but I can't add `zap.DynamicLevel().SetLevel(zap.WarnLevel)` as an option.

**Comment by anuptalwalkar** [comment]
Nevermind actually. I understand now. I have updated my changes as well https://github.com/uber-go/fx/pull/81

**Pull request merged**